## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import pandas as pd 

2022-11-21 18:17:01.401465: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [2]:
#  Import and read the charity_data.csv.
application_df = pd.read_csv("../Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [3]:
application_df.columns

Index(['EIN', 'NAME', 'APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION',
       'USE_CASE', 'ORGANIZATION', 'STATUS', 'INCOME_AMT',
       'SPECIAL_CONSIDERATIONS', 'ASK_AMT', 'IS_SUCCESSFUL'],
      dtype='object')

In [4]:
application_df['IS_SUCCESSFUL'].value_counts()

1    18261
0    16038
Name: IS_SUCCESSFUL, dtype: int64

In [5]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
dropped_df = application_df.drop(['EIN', 'NAME'], axis = 1)
dropped_df.head(30)

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0               T10       Independent          C1000    ProductDev   
1                T3       Independent          C2000  Preservation   
2                T5  CompanySponsored          C3000    ProductDev   
3                T3  CompanySponsored          C2000  Preservation   
4                T3       Independent          C1000     Heathcare   
5                T3       Independent          C1200  Preservation   
6                T3       Independent          C1000  Preservation   
7                T3       Independent          C2000  Preservation   
8                T7       Independent          C1000    ProductDev   
9                T5  CompanySponsored          C3000    ProductDev   
10               T3       Independent          C1200  Preservation   
11               T3       Independent          C2000  Preservation   
12               T3  CompanySponsored          C1200  Preservation   
13               T3       Independent          C2700  Preservation   
14               T3       Independent          C1000  Preservation   
15               T3       Independent          C1200  Preservation   
16               T3       Independent          C2000  Preservation   
17               T3       Independent          C1000     Heathcare   
18               T3  CompanySponsored          C1000  Preservation   
19               T3  CompanySponsored          C1000  Preservation   
20               T4       Independent          C3000    ProductDev   
21               T3       Independent          C1200  Preservation   
22               T3       Independent          C7000  Preservation   
23               T4  CompanySponsored          C3000    ProductDev   
24               T3  CompanySponsored          C1200  Preservation   
25               T4       Independent          C3000    ProductDev   
26               T3  CompanySponsored          C1000  Preservation   
27               T3  CompanySponsored          C2000  Preservation   
28               T3  CompanySponsored          C2000  Preservation   
29               T6       Independent          C2000    ProductDev   

    ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0    Association       1              0                      N      5000   
1   Co-operative       1         1-9999                      N    108590   
2    Association       1              0                      N      5000   
3          Trust       1    10000-24999                      N      6692   
4          Trust       1  100000-499999                      N    142590   
5          Trust       1              0                      N      5000   
6          Trust       1  100000-499999                      N     31452   
7          Trust       1        10M-50M                      N   7508025   
8          Trust       1         1-9999                      N     94389   
9    Association       1              0                      N      5000   
10         Trust       1    25000-99999                      N     69656   
11         Trust       1  100000-499999                      N    165593   
12   Association       1              0                      N      5000   
13         Trust       1    25000-99999                      N      5301   
14         Trust       1              0                      N      5000   
15         Trust       1              0                      N      5000   
16         Trust       1              0                      N      5000   
17         Trust       1           50M+                      N  86380556   
18         Trust       1              0                      N      5000   
19         Trust       1  100000-499999                      N    376109   
20         Trust       1              0                      N      5000   
21         Trust       1              0                      N      5000   
22         Trust       1  100000-499999                      N    318296   
23   Ass

In [6]:
# Find null values
for column in dropped_df.columns:
    print(f"Column {column} has {dropped_df[column].isnull().sum()} null values")

Column APPLICATION_TYPE has 0 null values
Column AFFILIATION has 0 null values
Column CLASSIFICATION has 0 null values
Column USE_CASE has 0 null values
Column ORGANIZATION has 0 null values
Column STATUS has 0 null values
Column INCOME_AMT has 0 null values
Column SPECIAL_CONSIDERATIONS has 0 null values
Column ASK_AMT has 0 null values
Column IS_SUCCESSFUL has 0 null values


In [7]:
# Find duplicate entries
print(f"Duplicate entries: {dropped_df.duplicated().sum()}")

Duplicate entries: 24677


In [8]:
# Determine the number of unique values in each column.
dropped_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [9]:
dropped_df.shape

(34299, 10)

In [10]:
dropped_df.describe()

STATUS       ASK_AMT  IS_SUCCESSFUL
count  34299.000000  3.429900e+04   34299.000000
mean       0.999854  2.769199e+06       0.532406
std        0.012073  8.713045e+07       0.498956
min        0.000000  5.000000e+03       0.000000
25%        1.000000  5.000000e+03       0.000000
50%        1.000000  5.000000e+03       1.000000
75%        1.000000  7.742000e+03       1.000000
max        1.000000  8.597806e+09       1.000000

In [11]:
dropped_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34299 entries, 0 to 34298
Data columns (total 10 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   APPLICATION_TYPE        34299 non-null  object
 1   AFFILIATION             34299 non-null  object
 2   CLASSIFICATION          34299 non-null  object
 3   USE_CASE                34299 non-null  object
 4   ORGANIZATION            34299 non-null  object
 5   STATUS                  34299 non-null  int64 
 6   INCOME_AMT              34299 non-null  object
 7   SPECIAL_CONSIDERATIONS  34299 non-null  object
 8   ASK_AMT                 34299 non-null  int64 
 9   IS_SUCCESSFUL           34299 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 2.6+ MB


In [12]:
# Look at APPLICATION_TYPE value counts for binning
val_counts = dropped_df['APPLICATION_TYPE'].value_counts()
val_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [13]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

# Transform APPLICATION_TYPE
def application_types_to_replace(application_type):
    if application_type == "T3":
        return 1
    else:
        return 0
    
# Along with replace() and map(), this is another way to encode the gender column into numbers.
dropped_df["APPLICATION_TYPE"] = dropped_df["APPLICATION_TYPE"].apply(application_types_to_replace)
dropped_df.head(20)

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  0       Independent          C1000    ProductDev   
1                  1       Independent          C2000  Preservation   
2                  0  CompanySponsored          C3000    ProductDev   
3                  1  CompanySponsored          C2000  Preservation   
4                  1       Independent          C1000     Heathcare   
5                  1       Independent          C1200  Preservation   
6                  1       Independent          C1000  Preservation   
7                  1       Independent          C2000  Preservation   
8                  0       Independent          C1000    ProductDev   
9                  0  CompanySponsored          C3000    ProductDev   
10                 1       Independent          C1200  Preservation   
11                 1       Independent          C2000  Preservation   
12                 1  CompanySponsored          C1200  Preservation   
13                 1       Independent          C2700  Preservation   
14                 1       Independent          C1000  Preservation   
15                 1       Independent          C1200  Preservation   
16                 1       Independent          C2000  Preservation   
17                 1       Independent          C1000     Heathcare   
18                 1  CompanySponsored          C1000  Preservation   
19                 1  CompanySponsored          C1000  Preservation   

    ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0    Association       1              0                      N      5000   
1   Co-operative       1         1-9999                      N    108590   
2    Association       1              0                      N      5000   
3          Trust       1    10000-24999                      N      6692   
4          Trust       1  100000-499999                      N    142590   
5          Trust       1              0                      N      5000   
6          Trust       1  100000-499999                      N     31452   
7          Trust       1        10M-50M                      N   7508025   
8          Trust       1         1-9999                      N     94389   
9    Association       1              0                      N      5000   
10         Trust       1    25000-99999                      N     69656   
11         Trust       1  100000-499999                      N    165593   
12   Association       1              0                      N      5000   
13         Trust       1    25000-99999                      N      5301   
14         Trust       1              0                      N      5000   
15         Trust       1              0                      N      5000   
16         Trust       1              0                      N      5000   
17         Trust       1           50M+                      N  86380556   
18         Trust       1              0                      N      5000   
19         Trust       1  100000-499999                      N    376109   

    IS_SUCCESSFUL  
0               1  
1               1  
2               0  
3               1  
4               1  
5               1  
6               1  
7               1  
8               1  
9               0  
10              0  
11              0  
12              1  
13              1  
14              1  
15              1  
16              1  
17              0  
18              0  
19              0

In [14]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = list(val_counts [val_counts < 1000].index)

# Replace in dataframe
for app in application_types_to_replace:
    dropped_df['APPLICATION_TYPE'] = dropped_df['APPLICATION_TYPE'].replace(app,"Other")

In [15]:
# Check to make sure binning was successful
dropped_df["APPLICATION_TYPE"].value_counts()

1    27037
0     7262
Name: APPLICATION_TYPE, dtype: int64

In [16]:
# Look at CLASSIFICATION value counts for binning
clssification_value_counts = dropped_df['CLASSIFICATION'].value_counts()
clssification_value_counts 

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [17]:
# You may find it helpful to look at CLASSIFICATION value counts >1
clssification_value_counts[clssification_value_counts  > 1]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: CLASSIFICATION, dtype: int64

In [18]:
clssification_value_counts[clssification_value_counts  > 10]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
Name: CLASSIFICATION, dtype: int64

In [19]:
clssification_value_counts[clssification_value_counts < 1000]

C7000    777
C1700    287
C4000    194
C5000    116
C1270    114
        ... 
C4120      1
C8210      1
C2561      1
C4500      1
C2150      1
Name: CLASSIFICATION, Length: 66, dtype: int64

In [20]:
# Determine which values to replace if counts are leass than 10
clssification_to_replace = list(clssification_value_counts[clssification_value_counts > 10].index)

# Replace in dataframe
for cls in clssification_to_replace:
    dropped_df['CLASSIFICATION'] = dropped_df['CLASSIFICATION'].replace(cls,"Other")

In [21]:
# Check to make sure binning was successful
dropped_df['CLASSIFICATION'].value_counts()

Other    34181
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C4100        6
C2400        6
C1720        6
C1257        5
C1600        5
C0           3
C2710        3
C1260        3
C3200        2
C1234        2
C1256        2
C1246        2
C1267        2
C2190        1
C4200        1
C6100        1
C5200        1
C1370        1
C2600        1
C1900        1
C1820        1
C1248        1
C1236        1
C1283        1
C2570        1
C1580        1
C3700        1
C2500        1
C1570        1
C2380        1
C1732        1
C1728        1
C2170        1
C4120        1
C8210        1
C4500        1
C1245        1
C2561        1
C2150        1
Name: CLASSIFICATION, dtype: int64

In [22]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies_df = pd.get_dummies(dropped_df)
dummies_df.head(30)

APPLICATION_TYPE  STATUS   ASK_AMT  IS_SUCCESSFUL  \
0                  0       1      5000              1   
1                  1       1    108590              1   
2                  0       1      5000              0   
3                  1       1      6692              1   
4                  1       1    142590              1   
5                  1       1      5000              1   
6                  1       1     31452              1   
7                  1       1   7508025              1   
8                  0       1     94389              1   
9                  0       1      5000              0   
10                 1       1     69656              0   
11                 1       1    165593              0   
12                 1       1      5000              1   
13                 1       1      5301              1   
14                 1       1      5000              1   
15                 1       1      5000              1   
16                 1       1      5000              1   
17                 1       1  86380556              0   
18                 1       1      5000              0   
19                 1       1    376109              0   
20                 0       1      5000              1   
21                 1       1      5000              1   
22                 1       1    318296              0   
23                 0       1      5000              0   
24                 1       1      5000              0   
25                 0       1      5000              1   
26                 1       1      5000              1   
27                 1       1     44004              1   
28                 1       1      5000              0   
29                 0       1      5000              1   

    AFFILIATION_CompanySponsored  AFFILIATION_Family/Parent  \
0                              0                          0   
1                              0                          0   
2                              1                          0   
3                              1                          0   
4                              0                          0   
5                              0                          0   
6                              0                          0   
7                              0                          0   
8                              0                          0   
9                              1                          0   
10                             0                          0   
11                             0                          0   
12                             1                          0   
13                             0                          0   
14                             0                          0   
15                             0                          0   
16                             0                          0   
17                             0                          0   
18                             1                          0   
19                             1                          0   
20                             0                          0   
21                             0                          0   
22                             0                          0   
23                             1                          0   
24                             1                          0   
25                             0                          0   
26                             1                          0   
27                             1                          0   
28                             1                          0   
29                             0                          0   

    AFFILIATION_Independent  AFFILIATION_National  AFFILIATION_Other  \
0                         1                     0                  0   
1                         1                     0                  0   
2                         0                     0                 

In [23]:
# Split our preprocessed data into our features and target arrays
X = dummies_df.drop(["IS_SUCCESSFUL"], axis='columns').values
y = dummies_df["IS_SUCCESSFUL"].values
                    

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [24]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [25]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0]) 
hidden_nodes_layer1 = 8
hidden_nodes_layer2 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation ="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 600       
                                                                 
 dense_1 (Dense)             (None, 5)                 45        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 651
Trainable params: 651
Non-trainable params: 0
_________________________________________________________________


2022-11-21 18:17:08.073983: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.


In [26]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [27]:
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
804/804 [==============================] - 2s 1ms/step - loss: 0.6576 - accuracy: 0.6737
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6135 - accuracy: 0.6899
Epoch 3/100
804/804 [==============================] - 1s 2ms/step - loss: 0.6038 - accuracy: 0.6932
Epoch 4/100
804/804 [==============================] - 2s 2ms/step - loss: 0.6006 - accuracy: 0.7005
Epoch 5/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5987 - accuracy: 0.7021
Epoch 6/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5976 - accuracy: 0.7023
Epoch 7/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5970 - accuracy: 0.7028
Epoch 8/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5967 - accuracy: 0.7037
Epoch 9/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5963 - accuracy: 0.7039
Epoch 10/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5960 - accura

804/804 [==============================] - 1s 940us/step - loss: 0.5932 - accuracy: 0.7058
Epoch 82/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5933 - accuracy: 0.7058
Epoch 83/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5933 - accuracy: 0.7063
Epoch 84/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5932 - accuracy: 0.7058
Epoch 85/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5934 - accuracy: 0.7060
Epoch 86/100
804/804 [==============================] - 1s 2ms/step - loss: 0.5933 - accuracy: 0.7059
Epoch 87/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5931 - accuracy: 0.7058
Epoch 88/100
804/804 [==============================] - 1s 924us/step - loss: 0.5930 - accuracy: 0.7068
Epoch 89/100
804/804 [==============================] - 1s 961us/step - loss: 0.5933 - accuracy: 0.7062
Epoch 90/100
804/804 [==============================] - 1s 896us/step - loss: 0.5933 - ac

In [28]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.6013 - accuracy: 0.6960 - 314ms/epoch - 1ms/step
Loss: 0.6012909412384033, Accuracy: 0.6959766745567322


In [29]:
# Export our model to HDF5 file
nn.save_weights("AlphabetSoupCharity.h5")

# Optomize the  Model

In [30]:
# Adjust the input data to ensure that no variables or outliers are causing confusion in the model, such as:
# Dropping more or fewer columns: 'EIN', 'NAME', 'USE_CASE','SPECIAL_CONSIDERATIONS'
dropped_df = application_df.drop(['EIN', 'NAME', 'USE_CASE','SPECIAL_CONSIDERATIONS'], axis = 1)
dropped_df.head(30)

APPLICATION_TYPE       AFFILIATION CLASSIFICATION  ORGANIZATION  STATUS  \
0               T10       Independent          C1000   Association       1   
1                T3       Independent          C2000  Co-operative       1   
2                T5  CompanySponsored          C3000   Association       1   
3                T3  CompanySponsored          C2000         Trust       1   
4                T3       Independent          C1000         Trust       1   
5                T3       Independent          C1200         Trust       1   
6                T3       Independent          C1000         Trust       1   
7                T3       Independent          C2000         Trust       1   
8                T7       Independent          C1000         Trust       1   
9                T5  CompanySponsored          C3000   Association       1   
10               T3       Independent          C1200         Trust       1   
11               T3       Independent          C2000         Trust       1   
12               T3  CompanySponsored          C1200   Association       1   
13               T3       Independent          C2700         Trust       1   
14               T3       Independent          C1000         Trust       1   
15               T3       Independent          C1200         Trust       1   
16               T3       Independent          C2000         Trust       1   
17               T3       Independent          C1000         Trust       1   
18               T3  CompanySponsored          C1000         Trust       1   
19               T3  CompanySponsored          C1000         Trust       1   
20               T4       Independent          C3000         Trust       1   
21               T3       Independent          C1200         Trust       1   
22               T3       Independent          C7000         Trust       1   
23               T4  CompanySponsored          C3000   Association       1   
24               T3  CompanySponsored          C1200         Trust       1   
25               T4       Independent          C3000         Trust       1   
26               T3  CompanySponsored          C1000         Trust       1   
27               T3  CompanySponsored          C2000         Trust       1   
28               T3  CompanySponsored          C2000         Trust       1   
29               T6       Independent          C2000         Trust       1   

       INCOME_AMT   ASK_AMT  IS_SUCCESSFUL  
0               0      5000              1  
1          1-9999    108590              1  
2               0      5000              0  
3     10000-24999      6692              1  
4   100000-499999    142590              1  
5               0      5000              1  
6   100000-499999     31452              1  
7         10M-50M   7508025              1  
8          1-9999     94389              1  
9               0      5000              0  
10    25000-99999     69656              0  
11  100000-499999    165593              0  
12              0      5000              1  
13    25000-99999      5301              1  
14              0      5000              1  
15              0      5000              1  
16              0      5000              1  
17           50M+  86380556              0  
18              0      5000              0  
19  100000-499999    376109              0  
20              0      5000              1  
21              0      5000              1  
22  100000-499999    318296              0  
23              0      5000              0  
24              0      5000              0  
25              0      5000              1  
26              0      5000              1  
27    25000-99999     44004              1  
28              0      5000              0  
29              0      5000              1

In [31]:
# Look at APPLICATION_TYPE value counts for binning
val_counts = dropped_df['APPLICATION_TYPE'].value_counts()
val_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [32]:
# Creating more bins for rare occurrences in columns: Determine which values to replace if counts are less than 10
# Increasing or decreasing the number of values for each bin.
application_types_to_replace = list(val_counts [val_counts > 10].index)


# Replace in dataframe
for app in application_types_to_replace:
    dropped_df['APPLICATION_TYPE'] = dropped_df['APPLICATION_TYPE'].replace(cls,"Other")

In [33]:
dropped_df['APPLICATION_TYPE'].value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [34]:
# Look at CLASSIFICATION value counts for binning
clssification_value_counts = dropped_df['CLASSIFICATION'].value_counts()
clssification_value_counts 

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [35]:
clssification_value_counts[clssification_value_counts  > 100]

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
Name: CLASSIFICATION, dtype: int64

In [36]:
# Determine which values to replace if counts are less than 1000
clssification_to_replace = list(clssification_value_counts [clssification_value_counts > 100].index)

# Replace in dataframe
for cls in clssification_to_replace:
    dropped_df['CLASSIFICATION'] = dropped_df['CLASSIFICATION'].replace(cls,"Other")

In [37]:
# Check to make sure binning was successful
dropped_df['CLASSIFICATION'].value_counts()

Other    33630
C2800       95
C7100       75
C1300       58
C1280       50
         ...  
C4120        1
C2561        1
C1245        1
C8210        1
C2150        1
Name: CLASSIFICATION, Length: 61, dtype: int64

In [38]:
# Convert categorical data to numeric with `pd.get_dummies`
dummies_df = pd.get_dummies(dropped_df)
dummies_df.head(30)

STATUS   ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  \
0        1      5000              1                     1   
1        1    108590              1                     0   
2        1      5000              0                     0   
3        1      6692              1                     0   
4        1    142590              1                     0   
5        1      5000              1                     0   
6        1     31452              1                     0   
7        1   7508025              1                     0   
8        1     94389              1                     0   
9        1      5000              0                     0   
10       1     69656              0                     0   
11       1    165593              0                     0   
12       1      5000              1                     0   
13       1      5301              1                     0   
14       1      5000              1                     0   
15       1      5000              1                     0   
16       1      5000              1                     0   
17       1  86380556              0                     0   
18       1      5000              0                     0   
19       1    376109              0                     0   
20       1      5000              1                     0   
21       1      5000              1                     0   
22       1    318296              0                     0   
23       1      5000              0                     0   
24       1      5000              0                     0   
25       1      5000              1                     0   
26       1      5000              1                     0   
27       1     44004              1                     0   
28       1      5000              0                     0   
29       1      5000              1                     0   

    APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  \
0                      0                     0                     0   
1                      0                     0                     0   
2                      0                     0                     0   
3                      0                     0                     0   
4                      0                     0                     0   
5                      0                     0                     0   
6                      0                     0                     0   
7                      0                     0                     0   
8                      0                     0                     0   
9                      0                     0                     0   
10                     0                     0                     0   
11                     0                     0                     0   
12                     0                     0                     0   
13                     0                     0                     0   
14                     0                     0                     0   
15                     0                     0                     0   
16                     0                     0                     0   
17                     0                     0                     0   
18                     0                     0                     0   
19                     0                     0                     0   
20                     0                     0                     0   
21                     0                     0                     0   
22                     0                     0                     0   
23                     0                     0                     0   
24                     0                     0                     0   
25                     0                     0                     0   
26                     0                     0                     0   
27                     0                     0                     0   
28                     0

In [39]:
# Add more neurons to a hidden layer.
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0]) 
hidden_nodes_layer1 = 40
hidden_nodes_layer2 = 25

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation ="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 40)                3000      
                                                                 
 dense_4 (Dense)             (None, 25)                1025      
                                                                 
 dense_5 (Dense)             (None, 1)                 26        
                                                                 
Total params: 4,051
Trainable params: 4,051
Non-trainable params: 0
_________________________________________________________________


In [40]:
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [41]:
# Add or reduce the number of epochs to the training regimen.
# Compile the model
# Train the model
fit_model = nn.fit(X_train_scaled, y_train, epochs=150)

Epoch 1/150
804/804 [==============================] - 1s 1ms/step - loss: 0.6158 - accuracy: 0.6910
Epoch 2/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5982 - accuracy: 0.7020
Epoch 3/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5962 - accuracy: 0.7032
Epoch 4/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5956 - accuracy: 0.7039
Epoch 5/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5951 - accuracy: 0.7046
Epoch 6/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5949 - accuracy: 0.7040
Epoch 7/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5946 - accuracy: 0.7046
Epoch 8/150
804/804 [==============================] - 1s 2ms/step - loss: 0.5938 - accuracy: 0.7058
Epoch 9/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5934 - accuracy: 0.7059
Epoch 10/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5937 - accura

804/804 [==============================] - 1s 1ms/step - loss: 0.5901 - accuracy: 0.7083
Epoch 82/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5900 - accuracy: 0.7079
Epoch 83/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5898 - accuracy: 0.7081
Epoch 84/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5902 - accuracy: 0.7084
Epoch 85/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5901 - accuracy: 0.7076
Epoch 86/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5900 - accuracy: 0.7083
Epoch 87/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5900 - accuracy: 0.7080
Epoch 88/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5899 - accuracy: 0.7084
Epoch 89/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5897 - accuracy: 0.7084
Epoch 90/150
804/804 [==============================] - 1s 1ms/step - loss: 0.5900 - accuracy: 

# Report on the Neural Network  Model

## Overview of the analysis: Explain the purpose of this analysis.

The purpose of this analysis is if nonprofit foundation Alphabet Soup can help it select the applicants for funding with the best chance of success in the ventures. I used the features in the provided dataset to create a binary classifier that could be able predict whether applicants will be successful if funded by Alphabet Soup.

## Results: Using bulleted lists and images to support your answers, address the following questions.

* Data Preprocessing

'IS_SUCCESSFUL', 'APPLICATION_TYPE', 'CLASSIFICATION' variables are the target(s) for the model as it is shown number of success for 'IS_SUCCESSFUL', and in order to narrow down the target value.
'APPLICATION_TYPE','CLASSIFICATION', 'STATUS', 'ASK_AMT', 'IS_SUCCESSFUL', 'AFFILIATION', 'INCOME_AMT', 'SPECIAL_CONSIDERATIONS' variables are the features for the model.
'EIN', 'NAME' variables were removed from the input data because they are neither targets nor features.



* Compiling, Training, and Evaluating the Model

I selected 40 and 25 hidden layers, and 150 epoch for my neural network model test because I'd like to test if/How many neurons, layers, and activation functions could/would affect the outcome.
I was not able to achieve the target model performance(75%)
I tried to increase numbers of hidden layeres, and epoch to increase model performance.

## Summary: 

After increasing the number of hidden layers, and epoch, the accuracy is decreased to 70.83% from 71.67%. My recommendation for future tests would be Cluster Analysis with K-means which I have shown a high accuracy through PCA(n_components=0.99) in the previous test. 